File: 0637-average-of-levels-in-binary-tree/0637-average-of-levels-in-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun averageOfLevels(root: TreeNode?): DoubleArray {
        val result = ArrayList<Double>()
        val currentNodes = mutableListOf<TreeNode>()
        if (root != null) currentNodes.add(root)
        while (!currentNodes.isEmpty()) {
            val size = currentNodes.size
            var sum = 0L
            for (i in 0..size-1) {
                val node = currentNodes.removeAt(0)
                if (node.left != null) currentNodes.add(node.left)
                if (node.right != null) currentNodes.add(node.right)
                sum += node.`val`
            }
            result.add(sum.toDouble() / size)
        }
        return result.toDoubleArray()
    }
}

File: 0637-average-of-levels-in-binary-tree/0637-average-of-levels-in-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function averageOfLevels(root: TreeNode | null): number[] {
    const result = []
    const currentNodes = [];
    if (root) currentNodes.push(root)
    while (currentNodes.length) {
        const size = currentNodes.length
        let sum = 0
        for (let i = 0; i < size; i++) {
            const node = currentNodes.shift()
            sum += node.val
            if (node.left) currentNodes.push(node.left)
            if (node.right) currentNodes.push(node.right)
        }
        result.push(sum / size)
    }
    return result
};